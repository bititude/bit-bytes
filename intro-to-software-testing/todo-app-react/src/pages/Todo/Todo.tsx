import { useState } from 'react';
import AddTodo from './components/AddTodo/AddTodo';
import TodoLists from './components/TodoLists/TodoList';
import { useTodo } from './hooks/useTodo';

const Todo: React.FC = () => {
  const [isFormVisible, setIsFormVisible] = useState(false);
  const { todos, isTodosLoading, error } = useTodo();
  return (
    <div className='p-5 h-auto w-11/12 md:w-[450px]  shadow-md rounded-md m-auto border-2 border-gray-400 flex flex-col items-center justify-start mt-[100px] mb-20'>
      <div className='flex justify-between items-center mb-7 w-full'>
        <h1 className='font-bold text-gray-700 text-2xl'>Your todos</h1>
        <button
          data-testid='toggle-form-btn'
          className={`rounded-full ${
            isFormVisible ? 'bg-red-500' : 'bg-green-500'
          }  w-12 h-12 flex items-start justify-center hover:scale-105 shadow-md hover:shadow-lg`}
          onClick={() => setIsFormVisible(!isFormVisible)}
        >
          <span className='text-4xl text-white font-bold '>
            {isFormVisible ? <>&#xd7;</> : <>&#43;</>}
          </span>
        </button>
      </div>
      {isFormVisible ? <AddTodo /> : null}
      {error ? <p>{error.error}</p> : <TodoLists todos={todos} loading={isTodosLoading} />}
    </div>
  );
};

export default Todo;
