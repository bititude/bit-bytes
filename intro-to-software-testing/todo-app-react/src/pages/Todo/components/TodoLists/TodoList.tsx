import { useMemo } from 'react';
import { Todo } from '../../../../types/Todo';
import TodoItem from '../TodoItem/TodoItem';

interface ITodoLists {
  todos: Todo[];
  loading: boolean;
}

const TodoList: React.FC<ITodoLists> = ({ todos, loading }) => {
  const pendingTodos = useMemo(() => todos.filter(todo => !todo.completed), [todos]);

  const completedTodos = useMemo(() => todos.filter(todo => todo.completed), [todos]);

  if (loading) {
    return <h1>Loading</h1>;
  }
  return (
    <>
      <div data-testid='pending-todos' className='w-full'>
        <h1 className='font-bold text-gray-700 text-xl self-start my-4'>Pending</h1>
        {pendingTodos.length === 0 ? (
          <p>No Pending todos ðŸ¥³</p>
        ) : (
          pendingTodos.map(todo => <TodoItem key={todo.id} todo={todo} />)
        )}
      </div>

      <h1 className='font-bold text-gray-700 text-xl self-start my-4'>Completed</h1>

      <div data-testid='completed-todos'>
        {completedTodos.length === 0 ? (
          <p>No Completed todos</p>
        ) : (
          completedTodos.map(todo => <TodoItem key={todo.id} todo={todo} />)
        )}
      </div>
    </>
  );
};

export default TodoList;
