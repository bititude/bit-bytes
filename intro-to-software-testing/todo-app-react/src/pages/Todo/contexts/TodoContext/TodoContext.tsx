import { createContext, ReactNode, useCallback, useEffect, useState } from 'react';
import { Todo } from '../../../../types/Todo';
import {
  createTodo,
  getTodos,
  markTodoAsComplete,
  removeTodo,
} from '../../../../services/api/todo/todo';

type TodoError = {
  type: 'ADD' | 'LOAD' | 'DELETE' | 'UPDATE';
  error: string;
};

type TodoContextType = {
  todos: Todo[];
  isAdding: boolean;
  addTodo: (todo: Todo) => void;
  deleteTodo: (id: string) => void;
  markAsCompleted: (id: string) => void;
  error: TodoError | null;
  isTodosLoading: boolean;
};

export const TodoContext = createContext<TodoContextType | undefined>(undefined);

export const TodoProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [todos, setTodos] = useState<Todo[]>([]);
  const [isAdding, setIsAdding] = useState(false);
  const [error, setError] = useState<TodoError | null>(null);
  const [isTodosLoading, setIsTodosLoading] = useState(false);

  const addTodo = useCallback(async (todo: Todo) => {
    try {
      setIsAdding(true);
      await createTodo(todo);
      setTodos(currentTodos => [todo, ...currentTodos]);
    } catch {
      setError({
        type: 'ADD',
        error: 'something went wrong creating the todo, please try again later',
      });
    } finally {
      setIsAdding(false);
    }
  }, []);

  const deleteTodo = useCallback(async (id: string) => {
    try {
      await removeTodo(id);
      setTodos(currentTodos => currentTodos.filter(todo => todo.id !== id));
    } catch {
      setError({
        type: 'DELETE',
        error: "sorry, we can't perform this actions right now, please try again later",
      });
    }
  }, []);

  const markAsCompleted = useCallback(async (id: string) => {
    try {
      await markTodoAsComplete(id);
      setTodos(currentTodos =>
        currentTodos.map(todo => (todo.id === id ? { ...todo, completed: true } : todo))
      );
    } catch {
      setError({
        type: 'UPDATE',
        error: "sorry, we can't perform this actions right now, please try again later",
      });
    }
  }, []);

  const loadTodos = useCallback(async () => {
    try {
      setIsTodosLoading(true);
      const data = await getTodos();
      setTodos(data);
    } catch {
      setError({
        type: 'LOAD',
        error: "sorry, we can't fetch your todos right now, please try again later",
      });
    } finally {
      setIsTodosLoading(false);
    }
  }, []);

  useEffect(() => {
    loadTodos();
  }, [loadTodos]);

  return (
    <TodoContext.Provider
      value={{ todos, addTodo, markAsCompleted, deleteTodo, error, isAdding, isTodosLoading }}
    >
      {children}
    </TodoContext.Provider>
  );
};
