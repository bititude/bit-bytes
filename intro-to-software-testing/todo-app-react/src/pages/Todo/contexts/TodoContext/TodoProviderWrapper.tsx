import { TodoProvider } from './TodoContext';

export const TodoProviderWrapper: React.FC = ({ children }) => (
  <TodoProvider>{children}</TodoProvider>
);
