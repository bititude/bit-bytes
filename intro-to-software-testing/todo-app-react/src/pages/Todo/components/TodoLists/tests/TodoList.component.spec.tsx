/**
 * Requirements
 * 1. should show a loading indicator on loading
 * 2. a message should be shown for no pending todos
 * 3. a message should be shown for no completed todos
 */

import { describe, expect, it } from 'vitest';
import TodoList from '../TodoList';
import { TodoProviderWrapper } from '../../../contexts/TodoContext/TodoProviderWrapper';
import { render, screen } from '@testing-library/react';
import { Todo } from '../../../../../types/Todo';
import { faker } from '@faker-js/faker';

describe('TodoList.tsx', () => {
  it('should show a loading indicator on loading', () => {
    const todos: Todo[] = [
      { id: faker.string.uuid(), completed: true, title: faker.lorem.lines(1) },
    ];
    render(<TodoList todos={todos} loading={true} />, { wrapper: TodoProviderWrapper });

    const loadingIndicator = screen.getByText('Loading');

    expect(loadingIndicator).toBeInTheDocument();
  });

  it('a message should be shown for no pending todos', () => {
    const todos: Todo[] = [];

    render(<TodoList todos={todos} loading={false} />, { wrapper: TodoProviderWrapper });

    const noPendingTodos = screen.getByText('No Pending todos ðŸ¥³');

    expect(noPendingTodos).toBeInTheDocument();
  });

  it('a message should be shown for no completed todos', () => {
    const todos: Todo[] = [];

    render(<TodoList todos={todos} loading={false} />, { wrapper: TodoProviderWrapper });

    const noPendingTodos = screen.getByText('No Completed todos');

    expect(noPendingTodos).toBeInTheDocument();
  });

  describe('should match snapshot', () => {
    it('should match default snapshot', () => {
      const todos: Todo[] = [{ id: faker.string.uuid(), completed: true, title: 'wash car' }];
      const { asFragment } = render(<TodoList todos={todos} loading={false} />, {
        wrapper: TodoProviderWrapper,
      });

      expect(asFragment()).toMatchSnapshot();
    });

    it('should match loading snapshot', () => {
      const todos: Todo[] = [{ id: faker.string.uuid(), completed: true, title: 'wash car' }];
      const { asFragment } = render(<TodoList todos={todos} loading={true} />, {
        wrapper: TodoProviderWrapper,
      });

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
