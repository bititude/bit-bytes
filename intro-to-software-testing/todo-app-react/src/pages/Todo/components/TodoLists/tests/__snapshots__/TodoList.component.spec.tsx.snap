// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TodoList.tsx > should match snapshot > should match default snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-full"
    data-testid="pending-todos"
  >
    <h1
      class="font-bold text-gray-700 text-xl self-start my-4"
    >
      Pending
    </h1>
    <p>
      No Pending todos ðŸ¥³
    </p>
  </div>
  <h1
    class="font-bold text-gray-700 text-xl self-start my-4"
  >
    Completed
  </h1>
  <div
    data-testid="completed-todos"
  >
    <div
      class="border-2 border-gray-400 py-2 px-4 font-bold rounded-md shadow-md mb-4 flex items-center justify-between w-full"
      data-testid="todo"
    >
      <p
        class="w-4/5 text-gray-500"
        data-testid="striked-todo"
      >
         
        <s
          class="decoration-2"
        >
          <i>
            wash car
          </i>
        </s>
      </p>
      <button
        class="rounded-full bg-red-500   w-6 h-6 flex items-start justify-center hover:scale-105 "
        data-testid="todo-delete-btn"
      >
        <span
          class=" text-white font-extrabold "
        >
          Ã—
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TodoList.tsx > should match snapshot > should match loading snapshot 1`] = `
<DocumentFragment>
  <h1>
    Loading
  </h1>
</DocumentFragment>
`;
